tools/shared/src/test/scala/org/scalajs/core/tools/linker/AnalyzerTest.scala

```diff
@@ -0,0 +1,240 @@
+    )
+
+    val analysis = computeAnalysis(classDefs, reqsFactory.classData("LA"))
+
+    assertHasError("CycleInInheritanceChain(LA, LB)", analysis) {
+      case CycleInInheritanceChain(List(a, b), `fromUnitTest`) =>
+        a.encodedName == "LA" && b.encodedName == "LB"
+    }
+  }
+
+  @Test
+  def missingClassDirect(): Unit = {
+    val analysis = computeAnalysis(Nil, reqsFactory.classData("LA"))
+
+    assertHasError("MissingClass(LA)", analysis) {
+      case MissingClass(a, `fromUnitTest`) =>
+        a.encodedName == "LA"
+    }
+  }
+
+  @Test
+  def missingClassParent(): Unit = {
+    val classDefs = Seq(
+        ClassDef(Ident("LA"), ClassKind.Class, Some(Ident("LB")), Nil, None,
+            Nil, Nil)(emptyOptHints)
+    )
+
+    val analysis = computeAnalysis(classDefs, reqsFactory.classData("LA"))
+
+    assertHasError("MissingClass(LB)", analysis) {
+      case MissingClass(b, FromClass(a)) =>
+        a.encodedName == "LA" && b.encodedName == "LB"
+    }
+  }
+
+  @Test
+  def notAModule(): Unit = {
+    val classDefs = Seq(
+        ClassDef(Ident("LA"), ClassKind.Class, Some(Ident(ObjectClass)), Nil,
+            None, List(trivialCtor("LA")), Nil)(emptyOptHints)
+    )
+
+    val analysis = computeAnalysis(classDefs, reqsFactory.accessModule("LA"))
+
+    assertHasError("NotAModule(LA)", analysis) {
+      case NotAModule(a, `fromUnitTest`) =>
+        a.encodedName == "LA"
+    }
+  }
+
+  @Test
+  def missingMethod(): Unit = {
+    val classDefs = Seq(
+        ClassDef(Ident("LA"), ClassKind.Class, Some(Ident(ObjectClass)), Nil,
+            None, List(trivialCtor("LA")), Nil)(emptyOptHints)
+    )
+
+    val analysis = computeAnalysis(classDefs,
+        reqsFactory.instantiateClass("LA", "init___") ++
+        reqsFactory.callMethod("LA", "foo__V"))
+
+    assertHasError("MissingMethod(LA.foo__V)", analysis) {
+      case MissingMethod(foo, `fromUnitTest`) =>
+        foo.owner.encodedName == "LA" && foo.encodedName == "foo__V"
+    }
+  }
+
+  @Test
+  def conflictingDefaultMethods(): Unit = {
+    val defaultMethodDef = MethodDef(static = false, Ident("foo__V"), Nil,
+        NoType, Some(Skip()))(emptyOptHints, None)
+    val classDefs = Seq(
+        ClassDef(Ident("LI1"), ClassKind.Interface, None, Nil,
+            None, List(defaultMethodDef), Nil)(emptyOptHints),
+        ClassDef(Ident("LI2"), ClassKind.Interface, None, Nil,
+            None, List(defaultMethodDef), Nil)(emptyOptHints),
+        ClassDef(Ident("LA"), ClassKind.Class, Some(Ident(ObjectClass)),
+            List(Ident("LI1"), Ident("LI2")), None, List(trivialCtor("LA")),
+            Nil)(emptyOptHints)
+    )
+
+    val analysis = computeAnalysis(classDefs,
+        reqsFactory.instantiateClass("LA", "init___") ++
+        reqsFactory.callMethod("LA", "foo__V"))
+
+    assertHasError("ConflictingDefaultMethods(LI1.foo__V, LI2.foo__V)", analysis) {
+      case ConflictingDefaultMethods(List(a, b), `fromAnalyzer`) =>
```

Comment:
Do we really want to impose the order of a, b in the test here?

Reply:
Not necessarily, but the analyzer is currently completely deterministic, so this won't spuriously fail. Should it change, we can revisit this test case to make it more permissive.

Alternatively, we could define `ConflictingDefaultMethods` with a `Set`.

Reply:
This is not about making the test flaky, but it's about making the test brittle: You do not want to test more than you expect from the specified implementation, so that when you change implementation details, you don't need to change the tests.

Otherwise you run into the risk of:
- Wrong changes to tests
- Maintenance cost

So I'm fine with any of the following:
- Define `ConflictingDefaultMethods` with a `Set`.
- Comparing the list of things in `ConflictingDefaultMethods` in an unordered fashion.
- Specifying the order (if such a specification is possible). I would prefer one of the other two though.